### PR TITLE
Process-message and message-start-event-subscription events have a key

### LINKED ISSUE #2805: Key of Message/WorkflowInstance-Subscription Records is not set
Every record (except initial command) has a key that identifies the entity. The key allows tracking the lifecycle of an entity (e.g. CREATED -> COMPLETE -> COMPLETE).

Currently, the key is not set for all Message- and WorkflowInstanceSubscription records. It is always `-1`. This makes it hard to track the lifecycle of the subscriptions.

AT:
* message- and workflow instance subscription records set the key for an entity

### PR DESCRIPTION
**Pull request by saig0**
## Description

* currently, all events of type process-message and message-start-event have no key set (i.e. it is always `-1`)
* in order to improve the visibility of the lifecycle of a subscription, generate a new key when creating it
* use this key to write all follow-up events of this subscription
* write additional events when a message start event subscription is deleted if the previous process contains more than one message start event
  * previously, the processor wrote just one deleted event for all subscriptions
  * changed the behavior to write one event per subscription  

## Related issues

closes #2805 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by npepinpe — APPROVED** [review]
:rocket: no blockers

- Can we have (or do we already?) a follow up issue to make use of this for indexing? At the very least we can simplify the subscription message API, but it may also help simplify other things in the state for indexing (though I'll admit I haven't fully thought about it, but I assume you did :slightly_smiling_face:)

File: engine/src/main/java/io/zeebe/engine/state/message/ProcessMessageSubscription.java [hunk 1]
```diff
@@ -22,9 +22,13 @@
   private final LongProperty commandSentTimeProp = new LongProperty("commandSentTime", 0);
   private final EnumProperty<State> stateProp =
       new EnumProperty<>("state", State.class, State.STATE_OPENING);
+  private final LongProperty keyProp = new LongProperty("key", -1L);
```
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/state/message/ProcessMessageSubscription.java:25]: Good catch :+1: I make the key mandatory.

**Comment by saig0** [comment]
Here is the follow-up issue to simplify the subscription endpoint: #3346. We could optimize the state access depending on the new endpoint.

**Pull request merged**